### ISSUE TITLE
is there a way to remote install gns3 with a customized version ? 

### ISSUE THREAD

**Issue by chunji08** [GNS3/gns3-server#2035]
I am working on the gns3 VMs that are built on a private Cloud Site.   

The general steps of my GNS3 VM setup is:  to build up a UBUNTU VM first. => then on top of the VM, I am running this script https://github.com/GNS3/gns3-server/blob/master/scripts/remote-install.sh, to have the GNS3 VM setup along with OPENVPN. 

But this script does not give me the option to choose GNS3 VM version, and it always used the latest one by default.  For example, if I was to have it set up now, the GNS3 VM would be of version 2.2.29. 

I am working on a huge project that lasts for 6-12 months, and I need to add extra GNS3 VM from time to time. If the list of GNS3 VMs are not of the same version, it would introduce lots of issue for me. 

So in short,  can we add the gns3 version option in this script. 

Thanks a lot 

Jack 

02/28/2022 message update: 
I guess the short answer is "not possible", as we always keep latest copy on apt download site.  

```
apt list -a gns3-server
Listing... Done
gns3-server/focal 2.2.31~focal1 amd64 [upgradable from: 2.2.30~focal1]
gns3-server/now 2.2.30~focal1 amd64 [installed,upgradable to: 2.2.31~focal1]

```

So any other options I may try to handle my case ?

**Comment by b-ehlers** [comment]
First of all a warning: I don't use the remote-install script, so all I'm talking about is theory.

The first idea I have, is to create an initial VM and then create a snapshot of that on the cloud. Later you can use that snapshot to create new VMs as identical clones. But of course, that requires that the cloud supports creating snapshots.

The second idea is to use PIP to install. As you already found out, the ubuntu package system doesn't support multiple versions of the same package, but PIP does. So replace the line `apt-get install -y gns3-server` in <https://github.com/GNS3/gns3-server/blob/22c16e3608a1daa0cbc7d41b6b3cf1914e442ba9/scripts/remote-install.sh#L157> by the following:

```
apt-get install -y python3 vpcs ubridge qemu-system-x86 qemu-kvm qemu-utils mtools cpulimit libvirt-daemon-system dynamips x11vnc xvfb
apt-get install -y python3-pip
pip3 install gns3-server==<version>
```

The first line installs all the dependencies of gns3-server, the second line installs PIP and the third one installs gns3-server by PIP.

As I said, this is just theory. If you want to use it, you have to make your own tests, be prepared to handle issues by yourself.

**Comment by chunji08** [comment]
Thanks for the tips.  I will give a try. 

In my previous project, I have followed this instruction to have set up a GNS3 Environment of 10+ VMs in Amazon AWS, https://orndor.com/2018/12/14/setup-a-gns3-server-on-aws-ec2/index.html. Everything worked exception this version "drift" issue.

**Comment by chunji08** [comment]
Here is the update. 
Following Bernhard's suggestion, I was able to have the gns3 installed of the older versions. But I am confused of the port that gns3server is using, should the port be 80 or 3080 ? 

- On my ubuntu GNS3 VM, it seems 80 is not  permitted for gns3 user. 
- But if I was to use 3080, I don't think I can have the UI and VM connected, here is the issue I have raised: https://github.com/GNS3/gns3-gui/issues/3308 !

Comments ? 

Chun

**Comment by chunji08** [comment]
Just want to give summarize to the issue that I have hit. 
1) login on my GNS3 VM host, I have found these info of the gns3server: 
```
ubuntu@us-880-2:~/gns3/PRO-US880-2$ which gns3server
/usr/local/bin/gns3server
ubuntu@us-880-2:~/gns3/PRO-US880-2$ head -3 /usr/local/bin/gns3server
#!/usr/bin/python3
# -*- coding: utf-8 -*-
import re
ubuntu@us-880-2:~/gns3/PRO-US880-2$ ls -l /usr/bin/python3
lrwxrwxrwx 1 root root 9 Mar 13  2020 /usr/bin/python3 -> python3.8

```
So, in order to have the gns3server to be able to use port 80, I need to run this command as the root:   
 `setcap CAP_NET_BIND_SERVICE=+eip /usr/bin/python3.8 `

2) This is no longer an issue, and I am able to have GNS3 UI to communicate with the GNS3 VM I have created. 

As of now, I think my initial issue has been resolved.
